### PR TITLE
bluetooth: Add host support for LE Connection Subrating feature

### PR DESCRIPTION
**Pull request by alexstanoev-nordic**
The PR adds support for LE Connection Subrating as defined in Core 5.4 Vol 6, Part B, Section 5.1.19.
As this is primarily a controller feature, the host support is mostly a wrapper around the relevant HCI commands.

Subrating also allows updating connection's peripheral latency and supervision timeout without doing a connection parameter update - this will update the stored parameters inside `bt_conn_le` in the link.

Also added a new Kconfig to cover host subrating support and added subrating functionality to the Bluetooth shell.

### CODE REVIEW

File: include/zephyr/bluetooth/conn.h [hunk 1]
```diff
@@ -196,6 +196,81 @@ struct bt_conn_le_data_len_param {
...
+	/** Maximum subrate factor. */
+	uint16_t subrate_max;
+	/** Maximum Peripheral latency in units of subrated connection intervals. */
+	uint16_t max_latency;
+	/** Minimum number of underlying connection events to remain active
+	 *  after a packet containing a Link Layer PDU with a non-zero Length
+	 *  field is sent or received.
+	 */
+	uint16_t continuation_number;
+	/** Connection Supervision timeout (N * 10 MS).
```
Comment (hermabe) [inline include/zephyr/bluetooth/conn.h:212]: Nit: milliseconds are `ms`. (Assuming the unit is not megasiemens)

File: include/zephyr/bluetooth/conn.h [hunk 2]
```diff
@@ -196,6 +196,81 @@ struct bt_conn_le_data_len_param {
...
+ * @param  __subrate_max Maximum subrate factor.
+ * @param  ___max_latency Maximum Peripheral latency in units of subrated
+ *                        connection intervals.
+ * @param  ___continuation_number Minimum number of underlying connection events
+ *                                to remain active after a packet containing a
+ *                                Link Layer PDU with a non-zero Length
+ *                                field is sent or received.
+ * @param  ___supervision_timeout Connection Supervision timeout (N * 10 MS).
+ */
+#define BT_CONN_LE_SUBRATE_PARAM_INIT(_subrate_min, _subrate_max, _max_latency, \
```
Comment (hermabe) [inline include/zephyr/bluetooth/conn.h:242]: Are these macros needed or should we require the application to create the struct manually like you do in the shell? A common complaint about zephyr and the bluetooth api is the heavy use of macros.

File: include/zephyr/bluetooth/conn.h [hunk 3]
```diff
@@ -1244,6 +1354,20 @@ struct bt_conn_cb {
...
+	 *
+	 *  This callback notifies the application that the subrating parameters
+	 *  of the connection have changed. The new parameters can be obtained
+	 *  using @ref bt_conn_get_info if status is BT_HCI_ERR_SUCCESS.
+	 *
+	 *  @param conn   Connection object.
+	 *  @param status HCI status from LE Subrate Change event.
+	 */
+	void (*subrate_changed)(struct bt_conn *conn,
+				const uint8_t status);
```
Comment (hermabe) [inline include/zephyr/bluetooth/conn.h:1368]: Did you consider giving a `bt_conn_le_subrating_info` in the callback as well? How useful is it to know that the subrating parameters changed without knowing the actual parameters? I expect most applications will want these parameters but I may be wrong.

File: subsys/bluetooth/host/conn.c [hunk 4]
```diff
@@ -3038,6 +3041,73 @@ int bt_conn_le_set_path_loss_mon_enable(struct bt_conn *conn, bool reporting_ena
...
+		return -ENOTSUP;
+	}
+
+	buf = bt_hci_cmd_create(BT_HCI_OP_LE_SET_DEFAULT_SUBRATE, sizeof(*cp));
+	if (!buf) {
+		return -ENOBUFS;
+	}
+
+	cp = net_buf_add(buf, sizeof(*cp));
+	cp->subrate_min = sys_cpu_to_le16(params->subrate_min);
```
Comment (hermabe) [inline subsys/bluetooth/host/conn.c:3079]: Should the host do any range checks on the parameters or are we ok with letting the controller do the checking for us?

File: subsys/bluetooth/host/hci_core.c [hunk 5]
```diff
@@ -1392,6 +1392,9 @@ static void update_conn(struct bt_conn *conn, const bt_addr_le_t *id_addr,
 	conn->le.data_len.rx_max_len = BT_GAP_DATA_LEN_DEFAULT;
 	conn->le.data_len.rx_max_time = BT_GAP_DATA_TIME_DEFAULT;
 #endif
+#if defined(CONFIG_BT_SUBRATING)
+	conn->le.subrate.factor = 1; /* No subrating. */
```
Comment (hermabe) [inline subsys/bluetooth/host/hci_core.c:1396]: `continuation_number` seems to be missing here.

File: include/zephyr/bluetooth/conn.h (hunk 2 above)
Comment (alexstanoev-nordic) [reply include/zephyr/bluetooth/conn.h:242]: My preference is for less macros as well, I added them to keep consistency with conn. update, PHY update and DLE in the same header, but happy to remove.

File: subsys/bluetooth/host/conn.c (hunk 4 above)
Comment (alexstanoev-nordic) [reply subsys/bluetooth/host/conn.c:3079]: No preference here, I saw not all commands do parameter validation, but doing it in the host means we can add some more helpful debug prints. Don't mind adding it if you think it's beneficial.

Comment (alexstanoev-nordic) [reply subsys/bluetooth/host/conn.c:3111]: Added range checks for `bt_conn_le_subrate_param`

File: include/zephyr/bluetooth/conn.h (hunk 3 above)
Comment (alexstanoev-nordic) [reply include/zephyr/bluetooth/conn.h]: Modified to pass the changed parameters to the callback now, is it the implementation you had in mind too?

File: subsys/bluetooth/Kconfig [hunk 6]
```diff
@@ -181,6 +181,13 @@ config BT_PATH_LOSS_MONITORING
 	  Enable support for LE Path Loss Monitoring feature that is defined in the
 	  Bluetooth Core specification, Version 5.4 | Vol 6, Part B, Section 4.6.32.
 
+config BT_SUBRATING
+	bool "LE Connection Subrating"
```
Comment (alwa-nordic) [inline subsys/bluetooth/Kconfig:185]: New APIs should start out experimental.

```suggestion
	bool "LE Connection Subrating"
	select EXPERIMENTAL
```

See https://docs.zephyrproject.org/latest/develop/api/api_lifecycle.html for more info.

File: include/zephyr/bluetooth/conn.h (hunk 3 above)
Comment (hermabe) [reply include/zephyr/bluetooth/conn.h]: Yes, something like this. Would be good to get feedback from some more people on the api questions

File: subsys/bluetooth/Kconfig (hunk 6 above)
Comment (alexstanoev-nordic) [reply subsys/bluetooth/Kconfig:185]: Thanks, will push together with any further review changes

**Review by hermabe — APPROVED** [review]

**Review by alwa-nordic — APPROVED** [review]

**Review by jhedberg — APPROVED** [review]

**Pull request merged**